src/Module.zig

```diff
@@ -2836,6 +2844,56 @@ fn wrapOptional(self: *Module, scope: *Scope, dest_type: Type, inst: *Inst) !*In
     return self.addUnOp(b, inst.src, dest_type, .wrap_optional, inst);
 }
 
+fn wrapErrorUnion(self: *Module, scope: *Scope, dest_type: Type, inst: *Inst) !*Inst {
+    // TODO deal with inferred error sets
+    const err_union = dest_type.castTag(.error_union).?;
+    if (inst.value()) |val| {
+        return self.constInst(scope, inst.src, .{
+            .ty = dest_type,
+            // creating a SubValue for the error_union payload
+            .val = try Value.Tag.error_union.create(
+                scope.arena(),
+                blk: {
+                    if (inst.ty.zigTypeTag() != .ErrorSet) {
+                        // .? because we know it has a value
+                        _ = try self.coerce(scope, err_union.data.payload, inst);
+                        break :blk val;
+                    } else {
+                        switch (err_union.data.error_set.tag()) {
+                            // TODO these should be moved to coerce with error union widening
```

Comment:
Why? This seems correct to me.

Reply:
Hmm, I guess `wrapErrorUnion` is a subset of coerce so its fine. If we see there is repeated code, we can always factor it out.